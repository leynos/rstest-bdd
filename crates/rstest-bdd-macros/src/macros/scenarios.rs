//! Implementation of the `scenarios!` macro.

use proc_macro::TokenStream;
use proc_macro2::{Span, TokenStream as TokenStream2};
use quote::{format_ident, quote};
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use crate::codegen::scenario::{ScenarioConfig, generate_scenario_code};
use crate::parsing::feature::{extract_scenario_steps, parse_and_load_feature};
use crate::utils::errors::error_to_tokens;

/// Sanitize a string so it may be used as a Rust identifier.
///
/// Non-alphanumeric characters are replaced with underscores and the result is
/// lowercased. Identifiers starting with a digit gain a leading underscore.
fn sanitize_ident(input: &str) -> String {
    let mut ident = String::new();
    for c in input.chars() {
        if c.is_ascii_alphanumeric() {
            ident.push(c.to_ascii_lowercase());
        } else {
            ident.push('_');
        }
    }
    if ident.is_empty() || ident.chars().next().is_some_and(|c| c.is_ascii_digit()) {
        ident.insert(0, '_');
    }
    ident
}

/// Recursively collect all `.feature` files under `base`.
fn collect_feature_files(base: &Path) -> std::io::Result<Vec<PathBuf>> {
    let mut files = Vec::new();
    for entry in fs::read_dir(base)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_dir() {
            files.extend(collect_feature_files(&path)?);
        } else if path.extension().is_some_and(|e| e == "feature") {
            files.push(path);
        }
    }
    Ok(files)
}

/// Parse the macro input and resolve the search directory.
///
/// Returns the Cargo manifest directory, the directory to search for feature
/// files, and the original directory string.
///
/// # Examples
///
/// ```rust,ignore
/// let input = quote::quote!("features").into();
/// std::env::set_var("CARGO_MANIFEST_DIR", "/tmp");
/// let (manifest, search, dir) = resolve_scenario_directory(input).unwrap();
/// assert_eq!(manifest, std::path::PathBuf::from("/tmp"));
/// assert_eq!(search, std::path::PathBuf::from("/tmp/features"));
/// assert_eq!(dir, "features");
/// ```
fn resolve_scenario_directory(
    input: TokenStream,
) -> Result<(PathBuf, PathBuf, String), TokenStream> {
    let dir_lit: syn::LitStr = match syn::parse(input) {
        Ok(lit) => lit,
        Err(err) => return Err(error_to_tokens(&err)),
    };
    let dir_value = dir_lit.value();
    let dir = PathBuf::from(&dir_value);

    let manifest_dir = if let Ok(v) = std::env::var("CARGO_MANIFEST_DIR") {
        PathBuf::from(v)
    } else {
        let err = syn::Error::new(
            Span::call_site(),
            "CARGO_MANIFEST_DIR is not set. This macro must run within Cargo.",
        );
        return Err(error_to_tokens(&err));
    };

    let search_dir = manifest_dir.join(&dir);
    Ok((manifest_dir, search_dir, dir_value))
}

/// Collect feature files from the given directory or return a compile error.
///
/// # Examples
///
/// ```rust,ignore
/// let paths = collect_and_validate_features(std::path::Path::new("./features"))
///     .unwrap();
/// assert!(!paths.is_empty());
/// ```
fn collect_and_validate_features(search_dir: &Path) -> Result<Vec<PathBuf>, TokenStream> {
    match collect_feature_files(search_dir) {
        Ok(v) => Ok(v),
        Err(err) => {
            let msg = format!("failed to read directory: {err}");
            let err = syn::Error::new(Span::call_site(), msg);
            Err(error_to_tokens(&err))
        }
    }
}

/// Generate the test code for every scenario inside a single feature file.
///
/// Deduplicates test names using `used_names`.
///
/// # Examples
///
/// ```rust,ignore
/// # use std::collections::HashSet;
/// let mut used = HashSet::new();
/// let tests = generate_test_for_scenario(
///     std::path::Path::new("alpha.feature"),
///     std::path::Path::new("/tmp"),
///     &mut used,
/// ).unwrap();
/// assert!(!tests.is_empty());
/// ```
fn generate_test_for_scenario(
    abs_path: &Path,
    manifest_dir: &Path,
    used_names: &mut HashSet<String>,
) -> Result<Vec<TokenStream2>, TokenStream> {
    let rel_path = abs_path
        .strip_prefix(manifest_dir)
        .map_or_else(|_| abs_path.to_path_buf(), Path::to_path_buf);

    let feature = parse_and_load_feature(rel_path.as_path())?;
    let feature_stem = sanitize_ident(
        rel_path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("feature"),
    );

    let mut tests = Vec::new();
    for (idx, _) in feature.scenarios.iter().enumerate() {
        let data = extract_scenario_steps(&feature, Some(idx))?;

        let base_name = format!("{}_{}", feature_stem, sanitize_ident(&data.name));
        let mut fn_name = base_name.clone();
        let mut counter = 1usize;
        while used_names.contains(&fn_name) {
            fn_name = format!("{base_name}_{counter}");
            counter += 1;
        }
        used_names.insert(fn_name.clone());
        let ident = format_ident!("{}", fn_name);

        let attrs: Vec<syn::Attribute> = Vec::new();
        let vis = syn::Visibility::Inherited;
        let sig: syn::Signature = syn::parse_quote! { fn #ident() };
        let block: syn::Block = syn::parse_quote!({});

        let feature_path = manifest_dir.join(&rel_path).display().to_string();

        let config = ScenarioConfig {
            attrs: &attrs,
            vis: &vis,
            sig: &sig,
            block: &block,
            feature_path,
            scenario_name: data.name,
            steps: data.steps,
            examples: data.examples,
        };
        let tokens = generate_scenario_code(config, std::iter::empty());
        tests.push(TokenStream2::from(tokens));
    }

    Ok(tests)
}

/// Wrap generated tests in a module named after the directory.
///
/// # Examples
///
/// ```rust,ignore
/// let module = create_scenarios_module("features", Vec::new());
/// assert!(module.to_string().contains("features"));
/// ```
fn create_scenarios_module(dir_value: &str, tests: &[TokenStream2]) -> TokenStream {
    let dir = PathBuf::from(dir_value);
    let module_ident = {
        let base = dir
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("scenarios");
        format_ident!("{}_scenarios", sanitize_ident(base))
    };

    let module_doc = format!("Scenarios auto-generated from `{dir_value}`.");

    TokenStream::from(quote! {
        #[doc = #module_doc]
        mod #module_ident {
            use super::*;
            #(#tests)*
        }
    })
}

/// Generate test modules for all scenarios within a directory of feature files.
pub(crate) fn scenarios(input: TokenStream) -> TokenStream {
    let (manifest_dir, search_dir, dir_value) = match resolve_scenario_directory(input) {
        Ok(v) => v,
        Err(err) => return err,
    };

    let feature_paths = match collect_and_validate_features(&search_dir) {
        Ok(v) => v,
        Err(err) => return err,
    };

    let mut used_names = HashSet::new();
    let mut tests = Vec::new();
    for abs_path in feature_paths {
        match generate_test_for_scenario(abs_path.as_path(), &manifest_dir, &mut used_names) {
            Ok(mut t) => tests.append(&mut t),
            Err(err) => return err,
        }
    }

    create_scenarios_module(&dir_value, &tests)
}

#[cfg(test)]
mod tests {
    use super::sanitize_ident;

    #[test]
    fn sanitises_invalid_identifiers() {
        assert_eq!(sanitize_ident("Hello world!"), "hello_world_");
    }
}
