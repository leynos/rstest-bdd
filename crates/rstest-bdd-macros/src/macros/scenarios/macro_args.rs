//! Parses arguments supplied to the `scenarios!` macro.
//!
//! Accepts either a positional directory literal or the `dir = "..."` and
//! `path = "..."` named arguments alongside an optional `tags = "..."` filter.
//! The parser enforces that each input appears at most once, mirroring both
//! accepted spellings in duplicate and missing-argument diagnostics so users
//! immediately see which synonym needs adjusting.

use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use syn::token::Comma;
use syn::LitStr;

pub(super) struct ScenariosArgs {
    pub(super) dir: LitStr,
    pub(super) tag_filter: Option<LitStr>,
}

enum ScenariosArg {
    Dir(LitStr),
    Tags(LitStr),
}

impl Parse for ScenariosArg {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        if input.peek(LitStr) {
            Ok(Self::Dir(input.parse()?))
        } else {
            let ident: syn::Ident = input.parse()?;
            input.parse::<syn::token::Eq>()?;
            if ident == "dir" || ident == "path" {
                Ok(Self::Dir(input.parse()?))
            } else if ident == "tags" {
                Ok(Self::Tags(input.parse()?))
            } else {
                Err(input.error("expected `dir`, `path`, or `tags`"))
            }
        }
    }
}

impl Parse for ScenariosArgs {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let args = Punctuated::<ScenariosArg, Comma>::parse_terminated(input)?;
        let mut dir = None;
        let mut tag_filter = None;

        for arg in args {
            match arg {
                ScenariosArg::Dir(lit) => {
                    if dir.is_some() {
                        return Err(input.error("duplicate `dir`/`path` argument"));
                    }
                    dir = Some(lit);
                }
                ScenariosArg::Tags(lit) => {
                    if tag_filter.is_some() {
                        return Err(input.error("duplicate `tags` argument"));
                    }
                    tag_filter = Some(lit);
                }
            }
        }

        let dir = dir.ok_or_else(|| input.error("`dir` (or `path`) is required"))?;

        Ok(Self { dir, tag_filter })
    }
}
