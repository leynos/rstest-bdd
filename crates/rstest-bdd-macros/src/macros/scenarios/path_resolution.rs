//! Canonicalises feature file paths for `scenarios!`, preferring capability-aware
//! resolution through `cap-std` before falling back to `std::fs` for escape
//! hatches.

use cap_std::ambient_authority;
use cap_std::fs::Dir;
use cap_std::AmbientAuthority;
use std::path::{Path, PathBuf};

fn canonicalize_absolute_path(
    path: &Path,
    authority: AmbientAuthority,
) -> std::io::Result<PathBuf> {
    let root = path
        .ancestors()
        .last()
        .unwrap_or_else(|| Path::new(std::path::MAIN_SEPARATOR_STR));
    let dir = Dir::open_ambient_dir(root, authority)?;
    let relative = path.strip_prefix(root).unwrap_or(path);
    let target = if relative.as_os_str().is_empty() {
        Path::new(".")
    } else {
        relative
    };
    let resolved = dir.canonicalize(target)?;
    if resolved.is_absolute() {
        Ok(resolved)
    } else {
        Ok(PathBuf::from(root).join(resolved))
    }
}

fn canonicalize_relative_path(
    path: &Path,
    authority: AmbientAuthority,
) -> std::io::Result<PathBuf> {
    let parent = path
        .parent()
        .filter(|p| !p.as_os_str().is_empty())
        .unwrap_or_else(|| Path::new("."));
    let dir = Dir::open_ambient_dir(parent, authority)?;
    let target = if parent == Path::new(".") {
        path
    } else {
        path.strip_prefix(parent).unwrap_or(path)
    };
    let resolved = dir.canonicalize(target)?;
    if resolved.is_absolute() {
        Ok(resolved)
    } else if parent == Path::new(".") {
        Ok(std::env::current_dir()?.join(resolved))
    } else {
        Ok(parent.to_path_buf().join(resolved))
    }
}

fn try_std_canonicalize_fallback(path: &Path, err: std::io::Error) -> std::io::Result<PathBuf> {
    if err.kind() == std::io::ErrorKind::PermissionDenied
        && err.to_string().contains("outside of the filesystem")
    {
        // cap-std denies canonicalising absolute symlinks that escape a
        // capability root. Falling back to std ensures we still support such
        // links whilst preferring capability-aware resolution for every other
        // case.
        std::fs::canonicalize(path)
    } else {
        Err(err)
    }
}

pub(super) fn canonicalize_path(path: &Path) -> std::io::Result<PathBuf> {
    let authority = ambient_authority();
    let attempt = if path.is_absolute() {
        canonicalize_absolute_path(path, authority)
    } else {
        canonicalize_relative_path(path, authority)
    };

    match attempt {
        Ok(resolved) => Ok(resolved),
        Err(err) => try_std_canonicalize_fallback(path, err),
    }
}
