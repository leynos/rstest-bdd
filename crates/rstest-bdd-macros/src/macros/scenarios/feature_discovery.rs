//! Discovers `.feature` files for the `scenarios!` macro without following
//! symlink targets. Delegates path canonicalisation to keep diagnostics
//! stable even when fixtures reside outside the workspace tree.

use std::path::{Path, PathBuf};

use walkdir::{DirEntry, WalkDir};

use super::path_resolution::canonicalize_path;

fn is_feature_file(path: &Path) -> bool {
    path.extension()
        .and_then(|e| e.to_str())
        .is_some_and(|ext| ext.eq_ignore_ascii_case("feature"))
}

/// Process a regular file entry, returning it if it's a feature file.
fn process_regular_file(path: PathBuf) -> Option<std::io::Result<PathBuf>> {
    if is_feature_file(&path) {
        Some(Ok(path))
    } else {
        None
    }
}

/// Process a special entry (e.g., symlink) by canonicalising and checking if
/// it's a feature file.
fn process_special_entry(original_path: PathBuf) -> Option<std::io::Result<PathBuf>> {
    match canonicalize_path(&original_path) {
        Ok(real_path) if real_path.is_file() && is_feature_file(&real_path) => {
            Some(Ok(original_path))
        }
        Ok(_) => None,
        Err(err) => Some(Err(err)),
    }
}

fn process_dir_entry(entry: DirEntry) -> Option<std::io::Result<PathBuf>> {
    let file_type = entry.file_type();
    if file_type.is_dir() {
        return None;
    }

    let path = entry.into_path();

    if file_type.is_file() {
        return process_regular_file(path);
    }

    process_special_entry(path)
}

fn convert_walkdir_error(err: walkdir::Error) -> Option<std::io::Error> {
    if err.loop_ancestor().is_some() {
        return None;
    }

    let err_str = err.to_string();
    Some(
        err.into_io_error()
            .unwrap_or_else(|| std::io::Error::other(err_str)),
    )
}

pub(super) fn collect_feature_files(base: &Path) -> std::io::Result<Vec<PathBuf>> {
    let mut files = Vec::new();

    for next in WalkDir::new(base).follow_links(false) {
        match next {
            Ok(entry) => {
                if let Some(result) = process_dir_entry(entry) {
                    files.push(result?);
                }
            }
            Err(err) => {
                if let Some(err) = convert_walkdir_error(err) {
                    return Err(err);
                }
            }
        }
    }

    files.sort();
    Ok(files)
}
