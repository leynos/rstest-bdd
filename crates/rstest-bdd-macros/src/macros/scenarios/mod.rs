//! Implementation of the `scenarios!` macro.

mod feature_discovery;
mod macro_args;
mod path_resolution;

use proc_macro::TokenStream;
use proc_macro2::{Span, TokenStream as TokenStream2};
use quote::{format_ident, quote};
use std::collections::HashSet;
use std::path::{Path, PathBuf};

use crate::codegen::scenario::{ScenarioConfig, generate_scenario_code};
use crate::parsing::feature::{ScenarioData, extract_scenario_steps, parse_and_load_feature};
use crate::parsing::tags::TagExpression;
use crate::utils::errors::{error_to_tokens, normalized_dir_read_error};
use crate::utils::fixtures::extract_function_fixtures;
use crate::utils::ident::sanitize_ident;

use self::feature_discovery::collect_feature_files;
use self::macro_args::ScenariosArgs;

struct ScenarioTestContext<'a> {
    feature_stem: &'a str,
    manifest_dir: &'a Path,
    rel_path: &'a Path,
}

fn dedupe_name(base: &str, used: &mut HashSet<String>) -> String {
    let mut name = base.to_string();
    let mut counter = 1usize;
    while used.contains(&name) {
        name = format!("{base}_{counter}");
        counter += 1;
    }
    used.insert(name.clone());
    name
}

struct TagFilter {
    expr: TagExpression,
    span: Span,
    raw: String,
}

fn generate_scenario_test(
    ctx: &ScenarioTestContext<'_>,
    used_names: &mut HashSet<String>,
    data: ScenarioData,
) -> TokenStream2 {
    let ScenarioData {
        name,
        steps,
        examples,
        ..
    } = data;
    let base_name = format!("{}_{}", ctx.feature_stem, sanitize_ident(&name));
    let fn_name = dedupe_name(&base_name, used_names);
    let fn_ident = format_ident!("{}", fn_name);

    let attrs: Vec<syn::Attribute> = Vec::new();
    let vis = syn::Visibility::Inherited;
    let mut sig: syn::Signature = examples.as_ref().map_or_else(
        || syn::parse_quote! { fn #fn_ident() },
        |ex| {
            let params = ex.headers.iter().map(|h| {
                let param_ident = format_ident!("{}", sanitize_ident(h));
                quote! { #[case] #param_ident: &'static str }
            });
            syn::parse_quote! { fn #fn_ident( #(#params),* ) }
        },
    );
    let Ok((_args, ctx_inserts)) = extract_function_fixtures(&mut sig) else {
        unreachable!("generated scenario signature must bind fixtures");
    };
    let block: syn::Block = syn::parse_quote!({});

    let feature_path = ctx.manifest_dir.join(ctx.rel_path).display().to_string();

    let config = ScenarioConfig {
        attrs: &attrs,
        vis: &vis,
        sig: &sig,
        block: &block,
        feature_path,
        scenario_name: name,
        steps,
        examples,
    };
    TokenStream2::from(generate_scenario_code(config, ctx_inserts.into_iter()))
}

#[expect(
    clippy::single_match_else,
    clippy::option_if_let_else,
    reason = "explicit match clarifies control flow"
)]
fn resolve_manifest_directory() -> Result<PathBuf, TokenStream> {
    match std::env::var("CARGO_MANIFEST_DIR") {
        Ok(v) => Ok(PathBuf::from(v)),
        Err(_) => {
            let err = syn::Error::new(
                Span::call_site(),
                "CARGO_MANIFEST_DIR is not set. This macro must run within Cargo.",
            );
            Err(error_to_tokens(&err).into())
        }
    }
}

fn process_scenarios(
    feature: &gherkin::Feature,
    ctx: &ScenarioTestContext<'_>,
    used_names: &mut HashSet<String>,
    tag_filter: Option<&TagExpression>,
) -> (Vec<TokenStream2>, Vec<TokenStream2>) {
    let mut tests = Vec::new();
    let mut errors = Vec::new();

    for (idx, _) in feature.scenarios.iter().enumerate() {
        match extract_scenario_steps(feature, Some(idx)) {
            Ok(mut data) => {
                if tag_filter.is_none_or(|filter| data.filter_by_tags(filter)) {
                    tests.push(generate_scenario_test(ctx, used_names, data));
                }
            }
            Err(err) => errors.push(err),
        }
    }

    (tests, errors)
}

fn process_feature_file(
    abs_path: &Path,
    manifest_dir: &Path,
    used_names: &mut HashSet<String>,
    tag_filter: Option<&TagExpression>,
) -> (Vec<TokenStream2>, Vec<TokenStream2>) {
    let rel_path = abs_path
        .strip_prefix(manifest_dir)
        .map_or_else(|_| abs_path.to_path_buf(), Path::to_path_buf);

    let feature = match parse_and_load_feature(rel_path.as_path()) {
        Ok(feature) => feature,
        Err(err) => return (Vec::new(), vec![err]),
    };

    let feature_stem = sanitize_ident(
        rel_path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("feature"),
    );
    let ctx = ScenarioTestContext {
        feature_stem: &feature_stem,
        manifest_dir,
        rel_path: &rel_path,
    };

    process_scenarios(&feature, &ctx, used_names, tag_filter)
}

fn generate_tests_from_features(
    feature_paths: Vec<PathBuf>,
    manifest_dir: &Path,
    tag_filter: Option<&TagExpression>,
) -> (Vec<TokenStream2>, Vec<TokenStream2>) {
    let mut used_names = HashSet::new();
    let mut tests = Vec::new();
    let mut errors = Vec::new();

    for abs_path in feature_paths {
        let (mut t, mut errs) = process_feature_file(
            abs_path.as_path(),
            manifest_dir,
            &mut used_names,
            tag_filter,
        );
        tests.append(&mut t);
        errors.append(&mut errs);
    }

    (tests, errors)
}

fn parse_tag_filter(tag_lit: Option<syn::LitStr>) -> Result<Option<TagFilter>, TokenStream> {
    tag_lit.map_or_else(
        || Ok(None),
        |lit| match TagExpression::parse(&lit.value()) {
            Ok(expr) => Ok(Some(TagFilter {
                expr,
                span: lit.span(),
                raw: lit.value(),
            })),
            Err(err) => {
                let syn_err = syn::Error::new(lit.span(), err.to_string());
                Err(error_to_tokens(&syn_err).into())
            }
        },
    )
}

fn check_empty_results(
    tests: &[TokenStream2],
    errors: &mut Vec<TokenStream2>,
    tag_filter: Option<&TagFilter>,
) {
    if tests.is_empty() && errors.is_empty() {
        if let Some(filter) = tag_filter {
            let err = syn::Error::new(
                filter.span,
                format!("no scenarios matched tag expression `{}`", filter.raw),
            );
            errors.push(error_to_tokens(&err));
        }
    }
}

pub(crate) fn scenarios(input: TokenStream) -> TokenStream {
    let ScenariosArgs {
        dir: dir_lit,
        tag_filter: tag_lit,
    } = syn::parse_macro_input!(input as ScenariosArgs);
    let dir = PathBuf::from(dir_lit.value());

    let tag_filter = match parse_tag_filter(tag_lit) {
        Ok(filter) => filter,
        Err(err_tokens) => return err_tokens,
    };

    let manifest_dir = match resolve_manifest_directory() {
        Ok(dir) => dir,
        Err(err_tokens) => return err_tokens,
    };

    let search_dir = manifest_dir.join(&dir);
    let feature_paths = match collect_feature_files(&search_dir) {
        Ok(paths) => paths,
        Err(err) => {
            let msg = normalized_dir_read_error(&search_dir, &err);
            let err = syn::Error::new(Span::call_site(), msg);
            return error_to_tokens(&err).into();
        }
    };

    let (tests, mut errors) = generate_tests_from_features(
        feature_paths,
        &manifest_dir,
        tag_filter.as_ref().map(|f| &f.expr),
    );

    check_empty_results(&tests, &mut errors, tag_filter.as_ref());

    let module_ident = {
        let base = dir
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("scenarios");
        format_ident!("{}_scenarios", sanitize_ident(base))
    };
    let module_doc = format!("Scenarios auto-generated from `{}`.", dir_lit.value());

    TokenStream::from(quote! {
        #[doc = #module_doc]
        mod #module_ident {
            use super::*;
            #(#tests)*
            #(#errors)*
        }
    })
}

#[cfg(test)]
mod tests {
    use super::dedupe_name;
    use std::collections::HashSet;

    #[test]
    fn deduplicates_duplicate_titles() {
        let mut used = HashSet::new();
        let first = dedupe_name("dup_same_name", &mut used);
        let second = dedupe_name("dup_same_name", &mut used);
        assert_eq!(first, "dup_same_name");
        assert_eq!(second, "dup_same_name_1");
    }

    #[cfg(unix)]
    mod unix {
        use super::super::feature_discovery::collect_feature_files;
        use std::fs;
        use std::io;
        use std::os::unix::fs::symlink;
        use std::path::Path;
        use tempfile::tempdir;

        #[test]
        fn collects_symlinked_feature_files_without_following_directory_loops() -> io::Result<()> {
            let temp = tempdir()?;
            let features_root = temp.path().join("features");
            fs::create_dir_all(features_root.join("nested"))?;

            let feature_path = features_root.join("nested/example.feature");
            fs::write(&feature_path, "Feature: Example\n")?;

            let symlink_path = features_root.join("symlink.feature");
            symlink(&feature_path, &symlink_path)?;

            let relative_symlink_path = features_root.join("relative_link.feature");
            symlink(Path::new("nested/example.feature"), &relative_symlink_path)?;

            let loop_dir = features_root.join("loop");
            symlink(&features_root, &loop_dir)?;

            let files = collect_feature_files(features_root.as_path())?;

            let mut expected = vec![feature_path, symlink_path, relative_symlink_path];
            expected.sort();
            assert_eq!(files, expected);

            Ok(())
        }
    }

    #[cfg(not(unix))]
    #[test]
    fn collects_symlinked_feature_files_without_following_directory_loops() {
        assert!(cfg!(not(unix)));
    }
}
