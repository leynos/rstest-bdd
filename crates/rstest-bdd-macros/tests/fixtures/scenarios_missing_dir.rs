use rstest_bdd_macros::scenarios;

scenarios!("tests/features/does_not_exist");

fn main() {}
