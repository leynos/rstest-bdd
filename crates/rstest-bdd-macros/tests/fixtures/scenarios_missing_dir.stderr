error: failed to read directory `$DIR/tests/features/does_not_exist`: No such file or directory (os error 2)
 --> tests/fixtures/scenarios_missing_dir.rs:3:1
  |
3 | scenarios!("tests/features/does_not_exist");
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `scenarios` (in Nightly builds, run with -Z macro-backtrace for more info)
